### ISSUE TITLE
azurerm_monitor_diagnostic_setting dynamic log/metric blocks produces inconsistent plan

### ISSUE THREAD

**Issue by liviudm** [terraform-providers/terraform-provider-azurerm#3900]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.12.4
+ provider.azurerm v1.31.0
```

### Affected Resource(s)

* `azurerm_monitor_diagnostic_setting`

### Terraform Configuration Files

```hcl
# azurerm_kubernetes_cluster, azurerm_storage_account and azurerm_log_analytics_workspace configuration omitted as they don't seem relevant

data "azurerm_monitor_diagnostic_categories" "aks" {
  resource_id = azurerm_kubernetes_cluster.aks.id
}

resource "azurerm_monitor_diagnostic_setting" "audit" {
  name                       = format("%s-aks-audit", var.project)
  target_resource_id         = azurerm_kubernetes_cluster.aks.id
  storage_account_id         = azurerm_storage_account.aks.id
  log_analytics_workspace_id = azurerm_log_analytics_workspace.aks.id

  dynamic "log" {
    iterator = log_category
    for_each = data.azurerm_monitor_diagnostic_categories.aks.logs

    content {
      category = log_category.value
      enabled  = true

      retention_policy {
        enabled = true
        days    = var.log_retention_days
      }
    }
  }

  dynamic "metric" {
    iterator = metric_category
    for_each = data.azurerm_monitor_diagnostic_categories.aks.metrics

    content {
      category = metric_category.value
      enabled  = true

      retention_policy {
        enabled = true
        days    = var.log_retention_days
      }
    }
  }
}
```

### Debug Output

```
Error: Provider produced inconsistent final plan

When expanding the plan for
module.tfkubev2_aks.azurerm_monitor_diagnostic_setting.audit to include new
values learned so far during apply, provider "azurerm" produced an invalid new
value for .log: block set length changed from 1 to 5.
```

### Expected Behavior

If the AKS cluster is created in the same `terraform apply` run, terraform fails with the above error. If the AKS cluster is already created, everything runs fine.

### Actual Behavior

Terraform produces an inconsistent plan and fails to provision `azurerm_monitor_diagnostic_setting`.

### Steps to Reproduce

1. Enable AKS audit logging [https://docs.microsoft.com/en-us/azure/aks/view-master-logs](https://docs.microsoft.com/en-us/azure/aks/view-master-logs)
2. `terraform apply`

**Comment by adubnik** [comment]
This issue does affect all diagnostic resources in azure provider when dynamic blocks are used and if metrics and/or logs for the resource are not defined and data source azurerm_monitor_diagnostic_categories is used to obtain valid logs and metrics for the resource. 

if used in one go in the dynamic block

```
dynamic "metric" {
    iterator = entry
    for_each = data.azurerm_monitor_diagnostic_categories.env.metrics

    content {
      category = entry.value
      enabled = true
      
      retention_policy {
        enabled = true
        days    = "${var.log_retention_days}"
      }
    }
  }
```

produces the issue 

```
Error: Provider produced inconsistent final plan

When expanding the plan for
module.azurerm_network_security_group.azurerm_monitor_diagnostic_setting.env
to include new values learned so far during apply, provider "azurerm" produced
an invalid new value for .metric: planned set element
cty.ObjectVal(map[string]cty.Value{"category":cty.UnknownVal(cty.String),
"enabled":cty.UnknownVal(cty.Bool),
"retention_policy":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"days":cty.UnknownVal(cty.Number),
"enabled":cty.UnknownVal(cty.Bool)})})}) does not correlate with any element
in actual.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

if resources + data are created without the diagnostic configuration and diagnostics are added on the later stage there is no issue. 

e.g. following is the output for the resource azurerm_network_security_group without creating the diagnostic resource...

logs = [
  "NetworkSecurityGroupEvent",
  "NetworkSecurityGroupRuleCounter",
]
metrics = []

**Comment by negeric** [comment]
It looks like this issue has been resolved with Terraform v0.12.9

**Comment by tombuildsstuff** [comment]
Closing since this @negeric says this has been fixed - if this is still an issue please let us know and we'll re-open this/take another look.

Thanks!
